src/Compilers/VisualBasic/Portable/Lowering/LambdaRewriter/LambdaRewriter.vb

```diff
@@ -907,11 +918,65 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return result
         End Function
 
+        Private Sub GetLambdaId(syntax As SyntaxNode, closureKind As ClosureKind, closureOrdinal As Integer, ByRef topLevelMethodId As MethodDebugId, ByRef lambdaOrdinal As Integer)
+            Debug.Assert(syntax IsNot Nothing)
+
+            Dim lambdaOrLambdaBodySyntax As SyntaxNode
+            Dim isLambdaBody As Boolean ' indicates whether a node needs mapping for lambda or its bodies
+            Dim syntaxNode = TryCast(syntax, LambdaExpressionSyntax)
+            If syntaxNode IsNot Nothing Then
+                ' lambda
+                isLambdaBody = True
+                lambdaOrLambdaBodySyntax = syntaxNode.SubOrFunctionHeader
+            ElseIf syntax.AncestorsAndSelf().OfType(Of QueryExpressionSyntax)().Any() Then
+                ' query lambdas
+                Select Case syntax.Kind
+                    Case SyntaxKind.GroupByClause,
+                         SyntaxKind.GroupJoinClause,
+                         SyntaxKind.FromClause,
+                         SyntaxKind.SimpleJoinClause,
+                         SyntaxKind.SelectClause
+                        isLambdaBody = False
+                    Case Else
+                        isLambdaBody = True
+                End Select
+                lambdaOrLambdaBodySyntax = syntax
```

Comment:
It feels like this code can benefit from a comment. At this point the intent is not clear and, therefore, there is no way to verify that the code does what it is supposed to do.